In [17]:
from googleapiclient.discovery import build

import pandas as pd
import seaborn as sns
from IPython.display import JSON

In [18]:
api_key = 'AIzaSyCYBMm0CU5lg49HR281zWs7i1P2p6pzJc8'
channel_id = 'UCX6b17PVsYBQ0ip5gyeme-Q'

In [19]:
youtube = build("youtube", "v3", developerKey=api_key)

# Stats

In [15]:
def get_channel_stats(youtube, channel_id):
        request = youtube.channels().list(
        part="snippet,contentDetails,statistics",
        id=channel_id)

        response = request.execute()

        data = dict(Channel_name = response['items'][0]['snippet']['title'],
                    Subscribers = response['items'][0]['statistics']['subscriberCount'],
                    views = response['items'][0]['statistics']['viewCount'],
                    Total_videos = response['items'][0]['statistics']['videoCount'],
                    playlist_id = response['items'][0]['contentDetails']['relatedPlaylists']['uploads']
        )
        return data


In [20]:
channel_statistics = get_channel_stats(youtube, channel_id)

channel_statistics

{'Channel_name': 'CrashCourse',
 'Subscribers': '15300000',
 'views': '1926030393',
 'Total_videos': '1514',
 'playlist_id': 'UUX6b17PVsYBQ0ip5gyeme-Q'}

{'Channel_name': 'CrashCourse',
 'Subscribers': '15300000',
 'views': '1925547276',
 'Total_videos': '1514',
 'playlist_id': 'UUX6b17PVsYBQ0ip5gyeme-Q'}


In [21]:
def get_video_ids(youtube, playlist_id):
        request = youtube.playlistItems().list(
        part="contentDetails",
        playlistId=playlist_id,
        maxResults = 50)

        response = request.execute()


        video_ids=[]
        for i in range (len(response['items'])):
                video_ids.append(response['items'][i]['contentDetails']['videoId'])


        next_page_token = response.get('nextPageToken')
        more_pages = True

        while more_pages:
                if next_page_token is None: #we are trying to get all the video ids, BUT bc the max return is 50, we are going through a loop of requests to get all video ids of the channel and storing it inside video_ids array:)
                        more_pages = False
                else:
                        request = youtube.playlistItems().list(
                                  part="contentDetails",
                                  playlistId=playlist_id,
                                  maxResults = 50,
                                  pageToken = next_page_token)
                        response = request.execute()

                        for i in range (len(response['items'])):
                                video_ids.append(response['items'][i]['contentDetails']['videoId'])
                        next_page_token = response.get('nextPageToken')

        return video_ids               


In [22]:
video_ids = get_video_ids(youtube, channel_statistics['playlist_id'])

JSON(video_ids)

<IPython.core.display.JSON object>

In [23]:
len(video_ids) #nice, we got all the video ids :)

1514

# Function to get video details

In [24]:
def get_video_details(youtube, video_ids):
    all_video_stats = []

    for i in range (0, len(video_ids), 50):
        request = youtube.videos().list(
                # part="snippet,statistics" original,
                  part="snippet,contentDetails,statistics", # هنا تبدأ خرابيط رنا

                id=','.join(video_ids[i:i+50]))
        response = request.execute()


        for video in response['items']:
            video_stats = dict(
                Title = video['snippet']['title'],
                ID = video['id'],
                Published_date = video['snippet']['publishedAt'],

                #Tags = video['snippet']['tags'], #+roy
                Tags = video['snippet'].get('tags', 0),

                #Views = video['statistics']['viewCount'],
                Views = video['statistics'].get('viewCount', 0),


                #Likes = video['statistics']['likeCount'],
                Likes = video['statistics'].get('likeCount', 0),

                #Comments = video['statistics']['commentCount'],
                Comments = video['statistics'].get('commentCount', 0),
                
                Duration = video['contentDetails']['duration'], #+roy
                Captoins = video['contentDetails']['caption'] #+roy
                )

            all_video_stats.append(video_stats)
    return all_video_stats
    


In [25]:
video_details = get_video_details(youtube, video_ids)

In [26]:
video_data = pd.DataFrame(video_details)
video_data

Title           ID  \
0     Black American History Arts & Culture | Compil...  bfH3fkIsc5U   
1     Why Are All Humans Unique? Meiosis: Crash Cour...  pj1oFx42d48   
2     Mitosis and the Cell Cycle: Crash Course Biolo...  skPOXcVvS5c   
3     Photosynthesis: The Original Solar Power: Cras...  -ZRsLhaukn8   
4     How do cells get their energy? (Electron Trans...  HeO3yagexTw   
...                                                 ...          ...   
1509    Water - Liquid Awesome: Crash Course Biology #2  HVT3Y3_gHGg   
1510  Indus Valley Civilization: Crash Course World ...  n7ndRwqJYDM   
1511       Carbon... SO SIMPLE: Crash Course Biology #1  QnQe0xW_JY4   
1512  The Agricultural Revolution: Crash Course Worl...  Yocja_N5s1I   
1513                               Crash Course Preview  FrlJl2UXRU0   

            Published_date                                               Tags  \
0     2024-02-16T16:30:06Z  [vlogbrothers, Crash Course, crashcourse, educ...   
1     2024-02-13T17:00:39Z  [vlogbrothers, Crash Course, crashcourse, educ...   
2     2024-02-06T17:00:44Z  [vlogbrothers, Crash Course, crashcourse, educ...   
3     2024-01-30T17:00:00Z  [vlogbrothers, Crash Course, crashcourse, educ...   
4     2024-01-23T17:00:13Z  [vlogbrothers, Crash Course, crashcourse, educ...   
...                    ...                                                ...   
1509  2012-02-06T22:10:10Z  [water, hydrogen, oxygen, molecule, covalent b...   
1510  2012-02-02T20:18:11Z  [John Green, Crashcourse, Documentary, Valley,...   
1511  2012-01-30T18:53:06Z  [biology, crashcourse, gilbert lewis, carbon, ...   
1512  2012-01-26T20:12:01Z  [John Green, history, agriculture, ancient, pr...   
1513  2011-12-02T20:07:14Z  [Crash, Course, Preview, vlogbrothers, world, ...   

         Views   Likes Comments    Duration Captoins  
0         9156     443       11  PT1H13M13S     true  
1        34217    1089       18    PT12M50S     true  
2        40245    1196       11    PT11M11S     true  
3        45988    1122       26     PT13M4S     true  
4        56870    1637       49     PT14M2S     true  
...        ...     ...      ...         ...      ...  
1509   5017999   52599     3819    PT11M17S     true  
1510   8207302   87815     8031     PT9M35S     true  
1511   7993263   80184     7788    PT11M57S     true  
1512  15894524  158062    12215    PT11M11S     true  
1513    430397   10204      850       PT39S     true  

[1514 rows x 9 columns]

In [27]:
# Save DataFrame to CSV file
video_data.to_csv('video_data.csv', index=False)

In [65]:
df = pd.read_csv('video_data.csv')
df.head()

Title           ID  \
0  Black American History Arts & Culture | Compil...  bfH3fkIsc5U   
1  Why Are All Humans Unique? Meiosis: Crash Cour...  pj1oFx42d48   
2  Mitosis and the Cell Cycle: Crash Course Biolo...  skPOXcVvS5c   
3  Photosynthesis: The Original Solar Power: Cras...  -ZRsLhaukn8   
4  How do cells get their energy? (Electron Trans...  HeO3yagexTw   

         Published_date                                               Tags  \
0  2024-02-16T16:30:06Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
1  2024-02-13T17:00:39Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
2  2024-02-06T17:00:44Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
3  2024-01-30T17:00:00Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
4  2024-01-23T17:00:13Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   

   Views  Likes  Comments    Duration  Captoins  
0   9156    443        11  PT1H13M13S      True  
1  34217   1089        18    PT12M50S      True  
2  40245   1196        11    PT11M11S      True  
3  45988   1122        26     PT13M4S      True  
4  56870   1637        49     PT14M2S      True

In [66]:
df.describe()

Views          Likes      Comments
count  1.514000e+03    1514.000000   1514.000000
mean   1.272184e+06   17799.279392   1028.836856
std    1.809005e+06   22416.910568   1846.079678
min    0.000000e+00       0.000000      0.000000
25%    1.739452e+05    4094.750000    118.000000
50%    5.391715e+05   10008.000000    432.500000
75%    1.601956e+06   22743.250000   1093.000000
max    1.589452e+07  248676.000000  29626.000000

In [67]:

# Get the number of rows and columns
num_rows, num_cols = df.shape

print("Number of rows:", num_rows)
print("Number of columns:", num_cols)

Number of rows: 1514
Number of columns: 9


In [68]:
column_types = df.dtypes

print("Types of columns:")
print(column_types)

Types of columns:
Title             object
ID                object
Published_date    object
Tags              object
Views              int64
Likes              int64
Comments           int64
Duration          object
Captoins            bool
dtype: object


In [69]:
# Check for null values
nulls_exist = df.isnull().any().any()

if nulls_exist:
    print("There are null values in the DataFrame.")
else:
    print("There are no null values in the DataFrame.")

There are no null values in the DataFrame.


In [72]:
df.Published_date.sort_values().value_counts()

Published_date
2011-12-02T20:07:14Z    1
2018-09-13T20:00:57Z    1
2018-10-02T19:43:42Z    1
2018-10-01T23:25:52Z    1
2018-09-28T19:59:32Z    1
                       ..
2016-01-16T15:00:01Z    1
2016-01-15T01:34:14Z    1
2016-01-13T18:18:54Z    1
2016-01-11T22:00:00Z    1
2024-02-16T16:30:06Z    1
Name: count, Length: 1514, dtype: int64

In [73]:
# Comments and likes per 1000 view ratio
df['likeRatio'] = df['Likes']/ df['Views'] * 1000
df['commentRatio'] = df['Comments']/ df['Views'] * 1000

df.head()

Title           ID  \
0  Black American History Arts & Culture | Compil...  bfH3fkIsc5U   
1  Why Are All Humans Unique? Meiosis: Crash Cour...  pj1oFx42d48   
2  Mitosis and the Cell Cycle: Crash Course Biolo...  skPOXcVvS5c   
3  Photosynthesis: The Original Solar Power: Cras...  -ZRsLhaukn8   
4  How do cells get their energy? (Electron Trans...  HeO3yagexTw   

         Published_date                                               Tags  \
0  2024-02-16T16:30:06Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
1  2024-02-13T17:00:39Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
2  2024-02-06T17:00:44Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
3  2024-01-30T17:00:00Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
4  2024-01-23T17:00:13Z  ['vlogbrothers', 'Crash Course', 'crashcourse'...   

   Views  Likes  Comments    Duration  Captoins  likeRatio  commentRatio  
0   9156    443        11  PT1H13M13S      True  48.383574      1.201398  
1  34217   1089        18    PT12M50S      True  31.826285      0.526054  
2  40245   1196        11    PT11M11S      True  29.717977      0.273326  
3  45988   1122        26     PT13M4S      True  24.397669      0.565365  
4  56870   1637        49     PT14M2S      True  28.784948      0.861614

In [81]:
from dateutil import parser


df['Published_date'] = pd.to_datetime(df['Published_date'])
df['Published_date'] = df['Published_date'].dt.strftime("%Y-%m-%d %H:%M:%S")

df.head()

Title           ID  \
0  Black American History Arts & Culture | Compil...  bfH3fkIsc5U   
1  Why Are All Humans Unique? Meiosis: Crash Cour...  pj1oFx42d48   
2  Mitosis and the Cell Cycle: Crash Course Biolo...  skPOXcVvS5c   
3  Photosynthesis: The Original Solar Power: Cras...  -ZRsLhaukn8   
4  How do cells get their energy? (Electron Trans...  HeO3yagexTw   

        Published_date                                               Tags  \
0  2024-02-16 16:30:06  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
1  2024-02-13 17:00:39  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
2  2024-02-06 17:00:44  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
3  2024-01-30 17:00:00  ['vlogbrothers', 'Crash Course', 'crashcourse'...   
4  2024-01-23 17:00:13  ['vlogbrothers', 'Crash Course', 'crashcourse'...   

   Views  Likes  Comments    Duration  Captoins  likeRatio  commentRatio  
0   9156    443        11  PT1H13M13S      True  48.383574      1.201398  
1  34217   1089        18    PT12M50S      True  31.826285      0.526054  
2  40245   1196        11    PT11M11S      True  29.717977      0.273326  
3  45988   1122        26     PT13M4S      True  24.397669      0.565365  
4  56870   1637        49     PT14M2S      True  28.784948      0.861614

AttributeError: Can only use .dt accessor with datetimelike values